runtime/src/main/java/org/corfudb/protocols/logprotocol/SMREntry.java

```diff
@@ -120,27 +122,34 @@ void deserializeBuffer(ByteBuf b, CorfuRuntime rt) {
         SMRMethod = new String(methodBytes);
         byte serializerId = b.readByte();
         byte numArguments = b.readByte();
-        Object[] arguments = new Object[numArguments];
+        List<Object> arguments = new ArrayList<>();
 
         if (!opaque) {
             serializerType = Serializers.getSerializer(serializerId);
         } else {
             this.serializerId = serializerId;
         }
 
+        List<ByteBuf> objBufPairs= new ArrayList<>();
         for (byte arg = 0; arg < numArguments; arg++) {
             int len = b.readInt();
             ByteBuf objBuf = b.slice(b.readerIndex(), len);
+            objBufPairs.add(objBuf);
+            b.skipBytes(len);
+        }
+
+        objBufPairs.parallelStream().forEachOrdered(objBuf -> {
             if (opaque) {
+                int len = objBuf.readableBytes();
                 byte[] argBytes = new byte[len];
                 objBuf.readBytes(argBytes);
-                arguments[arg] = argBytes;
+                arguments.add(argBytes);
             } else {
-                arguments[arg] = serializerType.deserialize(objBuf, rt);
+                arguments.add(serializerType.deserialize(objBuf, rt));
             }
-            b.skipBytes(len);
-        }
-        SMRArguments = arguments;
+        });
```

Comment:
I don't understand how this optimization works. I have a couple of questions. From the oracle docs:

> Note that you may lose the benefits of parallelism if you use operations like forEachOrdered with parallel streams

Since `forEachOrdered` is the last operation in the pipeline what is the point of using parallel stream, wouldn't `forEachOrdered` serialize the processing?

Most of the operations only have two arguments, so it's not clear this parallelization improves performance unless there are arguments that are collectios (i.e., putAll).

Also, `new ArrayList`, is not thread safe, see https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html

> Note that this implementation is not synchronized. If multiple threads access an ArrayList instance concurrently, and at least one of the threads modifies the list structurally, it must be synchronized externally.

Reply:
It looks like we lose parallelism if we use forEachOrdered https://stackoverflow.com/questions/47336825/benefit-of-using-foreachordered-with-parallel-streams